sympy/polys/matrices/linsolve.py

```diff
@@ -194,16 +199,30 @@ def _lin_eq2dict(a, symset):
                 terms = ti
                 terms_coeff = ci
             else:
-                raise PolyNonlinearError
-        coeff = Mul(*coeff_list)
+                # since ti is not null and we already have
+                # a term, this is a cross term
+                raise PolyNonlinearError(filldedent('''
+                    nonlinear cross-term: %s''' % a))
+        coeff = Mul._from_args(coeff_list)
         if terms is None:
             return coeff, {}
         else:
             terms = {sym: coeff * c for sym, c in terms.items()}
             return  coeff * terms_coeff, terms
-    elif a.is_Equality:
-        return _lin_eq2dict(a.lhs - a.rhs, symset)
-    elif not a.has_free(*symset):
+    if a.is_Equality:
+        (coeff, terms), (cR, tR) = [_lin_eq2dict(ai, symset)
+            for ai in a.args]
```

Comment:
Maybe equalities should be handled at a higher level before calling this function.

Reply:
I had already considered that. It is here because otherwise anyone using any of the linear coefficient extracting routines would have to duplicate this code to carefully make sure there is no linearity and then allow cancellation of linear terms.

Reply:
It could be handled by `_linear_eq_to_dict` for example.

The checking for zeros probably needs to be handled more broadly than just for `Equality`.